### PR TITLE
[ETH-FLOW v2] - Base PR - New ETH/WETH flow

### PR DESCRIPTION
**Pull request by W3stside**
Base PR for the new ETH/WETH flow

# Summary

#### _Currently only the new ETH/WETH wrapping, WETH/ETH unwrapping, and classic WETH trading functionalities are available in this PR._

This modifies/changes/improves the ETH/WETH flow in regards to:
1. Wrapping ETH > WETH
2. Unwrapping WETH > ETH
3. Swapping ETH > WETH <> ERC20 by obfuscating some of the wrap + approve logic necessary

### Native ETH flow is NOT available yet

# Testing
**Swapping ETH <> ERC20**
1. choose ETH as sell (or any native token)
2. choose an ERC20 as buy
3. Swap
Notes: you should get the entire wrapping/approving flow (if necessary)

**Wrapping ETH <> WETH**
1. choose ETH as sell (or any native token)
2. choose WETH as buy
3. Wrap

**Unwrapping WETH <> ETH**
1. choose WETH as sell (or any wrapped native token)
2. choose ETH as buy
3. Wrap

# Screenshots
### Non-expertMode
1. No more old eth flow!
image

2. Wrap ETH modal (if necessary)
image

3. Wrap ETH modal: pending transaction
 image

4. Approve WETH
image

5. Approve WETH: pending transaction
image

6. Confirmed meta-txs before swap
image

### expertMode
1. Wrap and approve both needed
image

2. Both operations pending 
image

3. Goes to swap confirmation
image

##### Note: to test, do as @elena-zh does, set a custom very low gasLimit (22k) and it will fail after some time
image

☝️ also applies the same to Approve failures and if BOTH fail during expertMode simultaneous tx signing

### Docs
Figma:
https://www.figma.com/proto/UGqD48GRUCiyav8z1xhhH2/%5BUX-RESEARCH%5D-%5D(W)ETH-Flows-(2022)?page-id=1%3A29&node-id=644%3A3899&viewport=976%2C-2470%2C0.06&scaling=min-zoom&starting-point-node-id=644%3A2927&show-proto-sidebar=1

Spec doc: https://docs.google.com/document/d/1AKZ7G3quQUw1PAbI43q64lMXHqg_BF9d9PQT4ciBaVQ/edit#

### Docs
Figma:
https://www.figma.com/proto/UGqD48GRUCiyav8z1xhhH2/%5BUX-RESEARCH%5D-%5D(W)ETH-Flows-(2022)?page-id=1%3A29&node-id=644%3A3899&viewport=976%2C-2470%2C0.06&scaling=min-zoom&starting-point-node-id=644%3A2927&show-proto-sidebar=1

Spec doc: https://docs.google.com/document/d/1AKZ7G3quQUw1PAbI43q64lMXHqg_BF9d9PQT4ciBaVQ/edit#

### CODE REVIEW

**Review by josojo — COMMENTED** [review]
Looks already super good.

By going over the PR I realized how rusty my ts/js is. But I tried my best ;)

File: src/custom/hooks/useApproveCallback/index.ts [hunk 1]
```diff
@@ -29,16 +39,35 @@ export function useApproveCallbackFromTrade({
   trade,
   allowedSlippage,
   amountToCheckAgainstAllowance,
-}: ApproveCallbackFromTradeParams) {
+  isNativeFlow,
+}: MakeOptional<ApproveCallbackFromTradeParams, 'openTransactionConfirmationModal' | 'closeModals'>): ApproveCallback
+export function useApproveCallbackFromTrade({
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/hooks/useApproveCallback/index.ts:44]: it is awkward :/

File: src/custom/hooks/useWrapCallback.ts [hunk 2]
```diff
@@ -221,8 +248,11 @@ export default function useWrapCallback(
 
     const isWrappingEther = inputCurrency.isNative && (isEthTradeOverride || weth.equals(outputCurrency))
     const isUnwrappingWeth = weth.equals(inputCurrency) && outputCurrency.isNative
+    // is an native currency trade but wrapped token has enough balance
+    const hasEnoughWrappedBalanceForSwap =
+      isEthTradeOverride && wrappedBalance && inputAmount && !wrappedBalance.lessThan(inputAmount)
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/hooks/useWrapCallback.ts:253]: well we also want equalTo but that lib doesn't have greaterThanOrEqualTo

File: src/custom/hooks/useApproveCallback/index.ts [hunk 3]
```diff
@@ -97,3 +127,70 @@ export function useApproveCallbackFromClaim({
...
+  return useApproveCallback({
+    openTransactionConfirmationModal,
+    closeModals,
+    spender: vaultRelayer,
+    amountToApprove: approveAmounts?.amountToApprove,
+    amountToCheckAgainstAllowance: approveAmounts?.amountToCheckAgainstAllowance,
+  })
+}
+
+export function useApproveNativeCallbackFromTrade({
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/hooks/useApproveCallback/index.ts:169]: i believe you're right and this should be deleted!

[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/hooks/useApproveCallback/index.ts:169]: #913

**Review by josojo — APPROVED** [review]
LGTM :)

File: src/custom/hooks/useApproveCallback/index.ts [hunk 4]
```diff
@@ -97,3 +127,70 @@ export function useApproveCallbackFromClaim({
...
+  trade,
+  allowedSlippage,
+  amountToCheckAgainstAllowance,
+}: ApproveCallbackFromTradeParams) {
+  const { chainId } = useActiveWeb3React()
+
+  const amountToApprove = useMemo(() => {
+    if (trade) {
+      const slippageForTrade = computeSlippageAdjustedAmounts(trade, allowedSlippage)
+      const token = slippageForTrade[Field.INPUT]?.wrapped
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/hooks/useApproveCallback/index.ts:181]: removing this code so resolving

**Comment by alfetopito** [comment]
What about the conflicts?
I assume you started this before the uni-merge. Bad luck 🤷

File: src/custom/state/swap/hooks.tsx [hunk 5]
```diff
@@ -493,8 +493,11 @@ export function useDetectNativeToken(input?: CurrencyWithAddress, output?: Curre
     // TODO: check the new native currency function
     const native = ETHER.onChain(chainId || DEFAULT_NETWORK_FOR_LISTS)
 
-    const [isNativeIn, isNativeOut] = [input?.currency?.isNative, output?.currency?.isNative]
-    const [isWrappedIn, isWrappedOut] = [input?.currency?.equals(wrappedToken), output?.currency?.equals(wrappedToken)]
+    const [isNativeIn, isNativeOut] = [!!input?.currency?.isNative, !!output?.currency?.isNative]
+    const [isWrappedIn, isWrappedOut] = [
+      !!input?.currency?.equals(wrappedToken),
+      !!output?.currency?.equals(wrappedToken),
+    ]
```
Comment (alfetopito) [inline src/custom/state/swap/hooks.tsx:500]: Was is failing somewhere?

**Review by alfetopito — APPROVED** [review]

File: src/custom/state/swap/hooks.tsx (hunk 5 above)
Comment (W3stside) [reply src/custom/state/swap/hooks.tsx:500]: no, just wasn't typed properly as `boolean | undefined` and later always checked as `!!isWrappedIn`

**Pull request merged**